pkg/mcp/client/client.go

```diff
@@ -247,138 +245,96 @@ func (c *Client) handleResponse(response *mcp.MeshConfigResponse) error {
-						log.Errorf("Error closing MCP stream after initial send failure: %v", err)
-					}
-					for {
-						if _, err := c.stream.Recv(); err != nil {
-							if err == io.EOF {
-								break
-							}
-							log.Errorf("Recv() error while waiting for MCP stream to close: %v", err)
-						}
-					}
-					retry = time.After(reestablishStreamDelay)
-					continue tryAgain
-				}
-				state.setVersion("")
-			}
-			return nil
-		}
-	}
+	return req
 }
 
 // Run starts the run loop for request and receiving configuration updates from
 // the server. This function blocks and should typically be run in a goroutine.
 // The client will continue to attempt to re-establish the stream with the server
 // indefinitely. The function exits when the provided context is canceled.
+//
+// See https://godoc.org/google.golang.org/grpc#ClientConn.NewStream
+// for rules to ensure stream resources are not leaked.
 func (c *Client) Run(ctx context.Context) {
-	responseC := make(chan *mcp.MeshConfigResponse)
-	var responseError error
-	receive := func() {
+	initRequests := make([]*mcp.MeshConfigRequest, 0, len(c.state))
+	for typeURL := range c.state {
+		initRequests = append(initRequests, &mcp.MeshConfigRequest{
+			Client:  c.clientInfo,
+			TypeUrl: typeURL,
+		})
+	}
+
+	for {
+		retry := time.After(time.Nanosecond)
 		for {
-			// On non-nil error the stream is closed or aborted. In either case,
-			// close the receive channel to signal to the for/select loop to re-establish
-			// the stream.
-			//
-			// (from https://godoc.org/google.golang.org/grpc#Stream)
-			//
-			// RecvMsg blocks until it receives a message or the stream is done. On
-			// client side, it returns io.EOF when the stream is done. On any other
-			// error, it aborts the stream and returns an RPC status.
-			response, err := c.stream.Recv()
-			if err != nil {
-				responseError = err
-				close(responseC)
+			select {
+			case <-ctx.Done():
 				return
+			case <-retry:
+			}
+
+			log.Info("(re)trying to establish new MCP stream")
+			var err error
+			if c.stream, err = c.client.StreamAggregatedResources(ctx); err == nil {
+				log.Info("New MCP stream created")
+				break
 			}
-			responseC <- response
+
+			log.Errorf("Failed to create a new MCP stream: %v", err)
+			retry = time.After(reestablishStreamDelay)
 		}
-	}
 
-	// establish the bidirectional grpc stream and start a dedicated
-	// goroutine to convert the synchronous stream.Recv() to a channel
-	// of responses.
-	if err := c.openStream(ctx); err != nil {
-		return
-	}
-	go receive()
+		var nextInitRequest int
 
-	for {
-		select {
-		case response, more := <-responseC:
-			if !more {
-				log.Errorf("Stream receive error: %v", responseError)
-
-				// re-establish the stream and start a new receive goroutine
-				// on any kind of error until the context is canceled.
-				if err := c.openStream(ctx); err != nil {
-					return
+		// The version and nonce information is reset for each new
```

Comment:
It is not clear why we have to stage the initial requests. What is the root problem?

Reply:
Updated the comments to hopefully clarify why we need to send initial requests.